src/HexagonOptimize.cpp

```diff
@@ -1605,6 +1616,293 @@ class OptimizeShuffles : public IRMutator {
 public:
     OptimizeShuffles(int lut_alignment) : lut_alignment(lut_alignment) {}
 };
+
+
+class VtmpyGenerator : public IRMutator {
+private:
+    using IRMutator::visit;
+
+    std::unordered_map<string, Expr> lets;
+    int vector_size;
+
+    // Return value of variable if expr is a variable.
+    // Else return the expr.
+    Expr getVal(Expr e) {
+        const Variable *op = e.as<Variable>();
+        if (op) {
+            if (lets.find(op->name) != lets.end())
+                return lets[op->name];
+            else 
+                return Expr();
+        } else {
+            return e;
+        }
+    }
+
+    void visit(const Let* op) {
+        lets[op->name] = op->value;
+        IRMutator::visit(op);
+        lets.erase(op->name);
+    }
+
+    void visit(const LetStmt *op) {
+        lets[op->name] = op->value;
+        IRMutator::visit(op);
+        lets.erase(op->name);
+    }
+
+    // Return the base of first vector if a and b are continuous vectors
+    Expr are_adjacent_vectors(const Expr &a, const Expr &b) {
+        Expr base1 = find_base(a);
```

Comment:
It seems to me that find_base walks through 'a' and finds the base. Consider writing this as an IRVisitor, not because that is typically how these things are written in Halide.

```
class BaseFinder : public IRVisitor {
private:
 Scope<Expr> lets;
 void visit(const Variable *op) {
}
 void visit(const Shuffle *op) {

}
public:
 BaseFinder(Scope<Expr> &lets) : lets(lets) {};
 Expr base;
};
Expr are_adjacent_vector(const Expr &a, const Expr &b) {
     BaseFinder f(lets);
     f.accept(&a);
     Expr base1 = f.base;
     f.accept(&a);
     Expr base2 = f.base;
}
```